File: backend/app/schemas/web_automation.py
"""
Web Automation Module Schemas
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID

from app.models.web_automation import (
    BrowserType, ExecutionMode, TestFlowStatus, 
    ExecutionRunStatus, StepStatus, HealingType, HealingStrategy
)


# Test Flow Schemas
class TestFlowCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=500)
    description: Optional[str] = None
    base_url: str = Field(default="")
    flow_json: Dict[str, Any] = Field(default_factory=dict)
    nodes: List[Dict[str, Any]] = Field(default_factory=list)
    edges: List[Dict[str, Any]] = Field(default_factory=list)
    viewport: Dict[str, Any] = Field(default_factory=dict)
    default_browser: BrowserType = BrowserType.CHROME
    default_mode: ExecutionMode = ExecutionMode.HEADED
    timeout: int = Field(default=30000, ge=1000, le=300000)
    retry_policy: str = Field(default="exponential")
    max_retries: int = Field(default=3, ge=0, le=10)
    healing_enabled: bool = True
    auto_update_selectors: bool = False
    healing_confidence_threshold: float = Field(default=0.75, ge=0.0, le=1.0)
    browser_options: Dict[str, Any] = Field(default_factory=dict)
    tags: List[str] = Field(default_factory=list)
    category: Optional[str] = None
    version: str = Field(default="1.0.0")


class TestFlowUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    status: Optional[TestFlowStatus] = None
    base_url: Optional[str] = None
    flow_json: Optional[Dict[str, Any]] = None
    nodes: Optional[List[Dict[str, Any]]] = None
    edges: Optional[List[Dict[str, Any]]] = None
    viewport: Optional[Dict[str, Any]] = None
    default_browser: Optional[BrowserType] = None
    default_mode: Optional[ExecutionMode] = None
    timeout: Optional[int] = Field(None, ge=1000, le=300000)
    retry_policy: Optional[str] = None
    max_retries: Optional[int] = Field(None, ge=0, le=10)
    healing_enabled: Optional[bool] = None
    auto_update_selectors: Optional[bool] = None
    healing_confidence_threshold: Optional[float] = Field(None, ge=0.0, le=1.0)
    browser_options: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    category: Optional[str] = None
    version: Optional[str] = None


class TestFlowResponse(BaseModel):
    id: UUID
    project_id: UUID
    organisation_id: UUID
    name: str
    description: Optional[str]
    status: TestFlowStatus
    base_url: str
    flow_json: Dict[str, Any]
    nodes: List[Dict[str, Any]]
    edges: List[Dict[str, Any]]
    viewport: Dict[str, Any]
    default_browser: BrowserType
    default_mode: ExecutionMode
    timeout: int
    retry_policy: str
    max_retries: int
    healing_enabled: bool
    auto_update_selectors: bool
    healing_confidence_threshold: float
    browser_options: Dict[str, Any]
    tags: List[str]
    category: Optional[str]
    version: str
    total_executions: int
    successful_executions: int
    failed_executions: int
    average_duration: int
    healing_success_rate: float
    created_at: datetime
    updated_at: Optional[datetime]
    created_by: Optional[UUID]
    last_executed_at: Optional[datetime]

    class Config:
        from_attributes = True


# Execution Schemas
class ExecutionRunCreate(BaseModel):
    browser_type: Optional[BrowserType] = None
    execution_mode: Optional[ExecutionMode] = None
    tags: List[str] = Field(default_factory=list)
    notes: Optional[str] = None
    trigger_source: str = Field(default="manual")


class ExecutionRunResponse(BaseModel):
    id: UUID
    test_flow_id: UUID
    project_id: UUID
    browser_type: BrowserType
    execution_mode: ExecutionMode
    status: ExecutionRunStatus
    total_steps: int
    passed_steps: int
    failed_steps: int
    skipped_steps: int
    healed_steps: int
    duration_ms: Optional[int]
    started_at: Optional[datetime]
    ended_at: Optional[datetime]
    execution_environment: Dict[str, Any]
    video_url: Optional[str]
    trace_url: Optional[str]
    screenshots_dir: Optional[str]
    error_message: Optional[str]
    error_stack: Optional[str]
    triggered_by: Optional[UUID]
    trigger_source: Optional[str]
    tags: List[str]
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class ExecutionRunDetailResponse(ExecutionRunResponse):
    step_results: List["StepResultResponse"] = []
    healing_events: List["HealingEventResponse"] = []


# Step Result Schemas
class StepResultResponse(BaseModel):
    id: UUID
    execution_run_id: UUID
    step_id: str
    step_name: Optional[str]
    step_type: str
    step_order: int
    status: StepStatus
    selector_used: Optional[Dict[str, Any]]
    action_details: Dict[str, Any]
    actual_result: Optional[str]
    expected_result: Optional[str]
    error_message: Optional[str]
    error_stack: Optional[str]
    duration_ms: Optional[int]
    retry_count: int
    screenshot_url: Optional[str]
    screenshot_before_url: Optional[str]
    screenshot_after_url: Optional[str]
    was_healed: bool
    healing_applied: Optional[Dict[str, Any]]
    console_logs: List[Dict[str, Any]]
    network_logs: List[Dict[str, Any]]
    started_at: Optional[datetime]
    ended_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


# Healing Event Schemas
class HealingEventResponse(BaseModel):
    id: UUID
    execution_run_id: UUID
    step_result_id: Optional[UUID]
    healing_type: HealingType
    strategy: HealingStrategy
    original_value: str
    healed_value: str
    step_id: str
    step_type: str
    failure_reason: Optional[str]
    success: bool
    confidence_score: Optional[float]
    retry_attempts: int
    ai_model: Optional[str]
    ai_reasoning: Optional[str]
    alternatives_tried: List[Dict[str, Any]]
    page_url: Optional[str]
    page_title: Optional[str]
    healing_duration_ms: Optional[int]
    recorded_at: datetime

    class Config:
        from_attributes = True


# Execution Request Schemas
class MultiBrowserExecutionRequest(BaseModel):
    browsers: List[BrowserType] = Field(..., min_items=1)
    execution_mode: ExecutionMode = ExecutionMode.HEADLESS
    parallel: bool = True
    tags: List[str] = Field(default_factory=list)
    notes: Optional[str] = None


class StopExecutionRequest(BaseModel):
    reason: Optional[str] = None


# Healing Suggestion Schemas
class LocatorHealingRequest(BaseModel):
    original_selector: str
    selector_strategy: str
    dom_snapshot: str
    page_url: str
    element_context: Dict[str, Any] = Field(default_factory=dict)
    failed_alternatives: List[Dict[str, Any]] = Field(default_factory=list)


class LocatorHealingSuggestion(BaseModel):
    suggested_selector: str
    strategy: HealingStrategy
    confidence_score: float
    reasoning: str
    alternatives: List[Dict[str, Any]] = Field(default_factory=list)


class AssertionHealingRequest(BaseModel):
    assertion_type: str
    expected_value: str
    actual_value: str
    selector: str
    tolerance: Optional[float] = None
    context: Dict[str, Any] = Field(default_factory=dict)


class AssertionHealingSuggestion(BaseModel):
    should_update: bool
    new_expected_value: Optional[str] = None
    confidence_score: float
    reasoning: str
    is_legitimate_change: bool


# Analytics Schemas
class HealingReportResponse(BaseModel):
    execution_run_id: UUID
    total_healings: int
    successful_healings: int
    failed_healings: int
    by_type: Dict[str, int]
    by_strategy: Dict[str, int]
    success_rate: float
    average_confidence: float
    timeline: List[Dict[str, Any]]


class TestFlowAnalytics(BaseModel):
    test_flow_id: UUID
    total_executions: int
    success_rate: float
    average_duration_ms: int
    healing_statistics: Dict[str, Any]
    browser_statistics: Dict[str, int]
    trend_data: List[Dict[str, Any]]


# Live Preview Schemas
class LiveUpdateMessage(BaseModel):
    type: str  # stepComplete, screenUpdate, console, locatorFound, assertion
    execution_run_id: UUID
    step_id: Optional[str] = None
    payload: Dict[str, Any]
    timestamp: datetime


# Locator Alternative Schemas
class LocatorAlternativeCreate(BaseModel):
    step_id: str
    element_identifier: str
    primary_selector: str
    primary_strategy: str
    alternatives: List[Dict[str, Any]] = Field(default_factory=list)
    element_attributes: Dict[str, Any] = Field(default_factory=dict)
    element_context: Dict[str, Any] = Field(default_factory=dict)
    page_url_pattern: Optional[str] = None


class LocatorAlternativeResponse(BaseModel):
    id: UUID
    test_flow_id: UUID
    step_id: str
    element_identifier: str
    primary_selector: str
    primary_strategy: str
    alternatives: List[Dict[str, Any]]
    success_count: int
    failure_count: int
    last_successful_selector: Optional[str]
    last_failed_at: Optional[datetime]
    element_attributes: Dict[str, Any]
    element_context: Dict[str, Any]
    page_url_pattern: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# Update forward references
ExecutionRunDetailResponse.model_rebuild()

File: frontend/lib/api/webAutomation.ts
import api from './index'

export interface TestFlow {
    id: string
    project_id: string
    organisation_id: string
    name: string
    description?: string
    status: 'draft' | 'active' | 'inactive' | 'archived'
    base_url: string
    flow_json: any
    nodes: any[]
    edges: any[]
    viewport: any
    default_browser: 'chrome' | 'firefox' | 'safari' | 'edge' | 'chromium'
    default_mode: 'headed' | 'headless'
    timeout: number
    retry_policy: string
    max_retries: number
    healing_enabled: boolean
    auto_update_selectors: boolean
    healing_confidence_threshold: number
    browser_options: any
    tags: string[]
    category?: string
    version: string
    total_executions: number
    successful_executions: number
    failed_executions: number
    average_duration: number
    healing_success_rate: number
    created_at: string
    updated_at: string
    last_executed_at?: string
}

export interface TestFlowCreate {
    name: string
    description?: string
    base_url?: string
    flow_json?: any
    nodes?: any[]
    edges?: any[]
    viewport?: any
    tags?: string[]
    category?: string
}

export interface TestFlowUpdate {
    name?: string
    description?: string
    status?: 'draft' | 'active' | 'inactive' | 'archived'
    base_url?: string
    flow_json?: any
    nodes?: any[]
    edges?: any[]
    viewport?: any
    default_browser?: string
    default_mode?: string
    timeout?: number
    healing_enabled?: boolean
    tags?: string[]
    category?: string
    steps?: any[]  // Test steps array
}

export const webAutomationApi = {
    listTestFlows: async (projectId: string) => {
        const response = await api.get<TestFlow[]>(`/api/v1/web-automation/projects/${projectId}/test-flows`)
        return response.data
    },

    getTestFlow: async (flowId: string) => {
        const response = await api.get<TestFlow>(`/api/v1/web-automation/test-flows/${flowId}`)
        return response.data
    },

    createTestFlow: async (projectId: string, data: TestFlowCreate) => {
        // Prepare default minimal flow data if not provided
        const payload = {
            flow_json: { nodes: [], edges: [], viewport: { x: 0, y: 0, zoom: 1 } },
            ...data,
        }
        const response = await api.post<TestFlow>(`/api/v1/web-automation/test-flows?project_id=${projectId}`, payload)
        return response.data
    },

    updateTestFlow: async (flowId: string, data: TestFlowUpdate) => {
        const response = await api.put<TestFlow>(`/api/v1/web-automation/test-flows/${flowId}`, data)
        return response.data
    },

    deleteTestFlow: async (flowId: string) => {
        await api.delete(`/api/v1/web-automation/test-flows/${flowId}`)
    },

    executeTestFlow: async (flowId: string, config?: any) => {
        const response = await api.post(`/api/v1/web-automation/test-flows/${flowId}/execute`, config || {})
        return response.data
    },

    startRecording: async (projectId: string, url: string) => {
        const response = await api.post(`/api/v1/web-automation/recorder/start`, { project_id: projectId, url })
        return response.data
    },

    stopRecording: async (projectId: string) => {
        const response = await api.post(`/api/v1/web-automation/recorder/stop`, { project_id: projectId })
        return response.data
    },

    // Browser Session Management - Live Browser Feature
    launchBrowserSession: async (sessionId: string, options?: {
        browserType?: string,
        device?: string,
        initialUrl?: string
    }) => {
        const params = new URLSearchParams({
            session_id: sessionId,
            browser_type: options?.browserType || 'chromium',
            device: options?.device || 'desktop_chrome',
            initial_url: options?.initialUrl || 'about:blank'
        })
        const response = await api.post(`/api/v1/web-automation/browser-sessions?${params}`)
        return response.data
    },

    stopBrowserSession: async (sessionId: string) => {
        const response = await api.delete(`/api/v1/web-automation/browser-sessions/${sessionId}`)
        return response.data
    },

    navigateBrowserSession: async (sessionId: string, url: string) => {
        const response = await api.post(`/api/v1/web-automation/browser-sessions/${sessionId}/navigate?url=${encodeURIComponent(url)}`)
        return response.data
    },

    highlightElement: async (sessionId: string, selector: string, duration?: number) => {
        const params = new URLSearchParams({
            selector,
            duration: String(duration || 2000)
        })
        const response = await api.post(`/api/v1/web-automation/browser-sessions/${sessionId}/highlight?${params}`)
        return response.data
    },

    getBrowserSessionState: async (sessionId: string) => {
        const response = await api.get(`/api/v1/web-automation/browser-sessions/${sessionId}`)
        return response.data
    },

    listBrowserSessions: async () => {
        const response = await api.get(`/api/v1/web-automation/browser-sessions`)
        return response.data
    },

    getDevicePresets: async () => {
        const response = await api.get<{ devices: Array<{ id: string, name: string, viewport: string, type: string }> }>(`/api/v1/web-automation/device-presets`)
        return response.data
    },

    // AI-Powered Step Generation
    generateStepsFromPrompt: async (prompt: string, context?: {
        currentUrl?: string,
        existingSteps?: Array<any>
    }): Promise<{
        success: boolean,
        steps: Array<{
            id: string,
            action: string,
            selector?: string,
            value?: string,
            url?: string,
            description?: string,
            [key: string]: any
        }>,
        explanation?: string,
        error?: string
    }> => {
        const response = await api.post(`/api/v1/web-automation/generate-steps`, {
            prompt,
            context
        })
        return response.data
    },

    // Execution Logs
    listProjectExecutions: async (projectId: string, options?: {
        status?: 'pending' | 'running' | 'completed' | 'failed' | 'stopped' | 'error',
        limit?: number,
        skip?: number
    }): Promise<ExecutionRun[]> => {
        const params = new URLSearchParams()
        if (options?.status) params.append('status', options.status)
        if (options?.limit) params.append('limit', String(options.limit))
        if (options?.skip) params.append('skip', String(options.skip))
        const query = params.toString() ? `?${params.toString()}` : ''
        const response = await api.get<ExecutionRun[]>(`/api/v1/web-automation/projects/${projectId}/executions${query}`)
        return response.data
    },

    listTestFlowExecutions: async (flowId: string, options?: {
        limit?: number,
        skip?: number
    }): Promise<ExecutionRun[]> => {
        const params = new URLSearchParams()
        if (options?.limit) params.append('limit', String(options.limit))
        if (options?.skip) params.append('skip', String(options.skip))
        const query = params.toString() ? `?${params.toString()}` : ''
        const response = await api.get<ExecutionRun[]>(`/api/v1/web-automation/test-flows/${flowId}/executions${query}`)
        return response.data
    },

    getExecutionRun: async (runId: string): Promise<ExecutionRunDetail> => {
        const response = await api.get<ExecutionRunDetail>(`/api/v1/web-automation/executions/${runId}`)
        return response.data
    },

    // Self-Heal Dashboard
    getSelfHealDashboard: async (projectId: string): Promise<SelfHealDashboard> => {
        const response = await api.get<SelfHealDashboard>(`/api/v1/web-automation/projects/${projectId}/self-heal/dashboard`)
        return response.data
    },

    // Apply healing fix
    applyHealingFix: async (stepId: string, suggestion: { value: string, type: string }): Promise<void> => {
        await api.post(`/api/v1/web-automation/healing/apply`, { step_id: stepId, suggestion })
    },

    // Artifacts
    listArtifacts: async (projectId: string, options?: {
        type?: 'screenshot' | 'video',
        page?: number,
        pageSize?: number
    }): Promise<ArtifactListResponse> => {
        const params = new URLSearchParams()
        if (options?.type) params.append('type', options.type)
        if (options?.page) params.append('page', String(options.page))
        if (options?.pageSize) params.append('page_size', String(options.pageSize))
        const query = params.toString() ? `?${params.toString()}` : ''
        const response = await api.get<ArtifactListResponse>(`/api/v1/projects/${projectId}/artifacts${query}`)
        return response.data
    },

    deleteArtifact: async (projectId: string, artifactId: string): Promise<void> => {
        await api.delete(`/api/v1/projects/${projectId}/artifacts/${artifactId}`)
    },

    getArtifactDownloadUrl: (projectId: string, artifactId: string): string => {
        return `/api/v1/projects/${projectId}/artifacts/${artifactId}/download`
    }
}

// Execution Types
export interface ExecutionRun {
    id: string
    test_flow_id: string
    project_id: string
    browser_type: 'chrome' | 'firefox' | 'safari' | 'edge' | 'chromium'
    execution_mode: 'headed' | 'headless'
    status: 'pending' | 'running' | 'completed' | 'failed' | 'stopped' | 'error'
    total_steps: number
    passed_steps: number
    failed_steps: number
    skipped_steps: number
    healed_steps: number
    duration_ms?: number
    started_at?: string
    ended_at?: string
    execution_environment: any
    video_url?: string
    trace_url?: string
    screenshots_dir?: string
    error_message?: string
    error_stack?: string
    triggered_by?: string
    trigger_source?: string
    tags: string[]
    notes?: string
    created_at: string
    test_flow_name?: string
}

export interface StepResult {
    id: string
    execution_run_id: string
    step_id: string
    step_name?: string
    step_type: string
    step_order: number
    status: 'pending' | 'running' | 'passed' | 'failed' | 'skipped' | 'healed'
    selector_used?: any
    action_details: any
    actual_result?: string
    expected_result?: string
    error_message?: string
    error_stack?: string
    duration_ms?: number
    retry_count: number
    screenshot_url?: string
    screenshot_before_url?: string
    screenshot_after_url?: string
    was_healed: boolean
    healing_applied?: any
    console_logs: any[]
    network_logs: any[]
    started_at?: string
    ended_at?: string
    created_at: string
}

export interface ExecutionRunDetail extends ExecutionRun {
    step_results: StepResult[]
    healing_events: any[]
}

export interface SelfHealDashboard {
    health_score: number
    total_tests: number
    issues_detected: number
    auto_healed_this_week: number
    detected_issues: SelfHealIssue[]
    repair_history: RepairHistoryItem[]
    config: {
        auto_apply_low_risk: boolean
        notify_on_issues: boolean
        visual_matching: boolean
        confidence_threshold: number
    }
}

export interface SelfHealIssue {
    id: string
    type: string
    test: string
    step: string
    status: string
    confidence: number
    old_locator: string
    error_message?: string
    suggestions: {
        id: string
        value: string
        confidence: number
        type: string
    }[]
}

export interface RepairHistoryItem {
    id: string
    date: string
    type: string
    test: string
    action: string
    success: boolean
}

// Artifact Types
export interface Artifact {
    id: string
    project_id: string
    execution_run_id?: string
    step_result_id?: string
    name: string
    type: 'screenshot' | 'video'
    file_path: string
    file_url?: string
    size_bytes?: number
    duration_ms?: number
    test_name?: string
    step_name?: string
    created_at: string
}

export interface ArtifactListResponse {
    items: Artifact[]
    total: number
    page: number
    page_size: number
    has_more: boolean
}

// ============================================
// Snippet Types
// ============================================
export interface SnippetParameter {
    name: string
    type: 'string' | 'number' | 'boolean' | 'selector'
    default?: string
    description?: string
}

export interface Snippet {
    id: string
    project_id: string
    organisation_id: string
    name: string
    description?: string
    parameters: SnippetParameter[]
    steps: any[]
    tags: string[]
    is_global: boolean
    version: string
    usage_count: number
    created_at: string
    updated_at?: string
    created_by?: string
}

export interface SnippetCreate {
    name: string
    description?: string
    parameters: SnippetParameter[]
    steps: any[]
    tags?: string[]
    is_global?: boolean
}

// Snippet API methods
export const snippetApi = {
    listSnippets: async (projectId: string, options?: { includeGlobal?: boolean, tag?: string, search?: string }) => {
        const params = new URLSearchParams({ project_id: projectId })
        if (options?.includeGlobal !== undefined) params.append('include_global', String(options.includeGlobal))
        if (options?.tag) params.append('tag', options.tag)
        if (options?.search) params.append('search', options.search)
        const response = await api.get<Snippet[]>(`/api/v1/snippets?${params}`)
        return response.data
    },

    getSnippet: async (snippetId: string) => {
        const response = await api.get<Snippet>(`/api/v1/snippets/${snippetId}`)
        return response.data
    },

    createSnippet: async (projectId: string, data: SnippetCreate) => {
        const response = await api.post<Snippet>(`/api/v1/snippets?project_id=${projectId}`, data)
        return response.data
    },

    updateSnippet: async (snippetId: string, data: Partial<SnippetCreate>) => {
        const response = await api.put<Snippet>(`/api/v1/snippets/${snippetId}`, data)
        return response.data
    },

    deleteSnippet: async (snippetId: string) => {
        await api.delete(`/api/v1/snippets/${snippetId}`)
    },

    createFromSteps: async (projectId: string, data: {
        name: string
        description?: string
        step_ids: string[]
        steps?: any[]
        parameters?: SnippetParameter[]
        tags?: string[]
        is_global?: boolean
    }) => {
        const response = await api.post<Snippet>(`/api/v1/snippets/from-steps?project_id=${projectId}`, data)
        return response.data
    },
}
